Fuse the input -> p2p buffer copy into one-shot all-reduce kernel when the input is small

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121213

This improves the gpt-fast llama2 70B 8xH100 (non-standard) TP benchmark from 86 tok/s to 88 tok/s.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

Commit:
Fuse the input -> p2p buffer copy into one-shot all-reduce kernel when the input is small

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/intra_node_comm.cu:
```diff
@@ -139,12 +139,21 @@ static __global__ void oneShotAllReduceKernel(
   const size_t stride = blockDim.x * gridDim.x * numelPerThread;
 
+  if (fuseInputCopy) {
+    for (size_t i = offset; i < N_aligned; i += stride) {
+      bf16x8 val;
+      streamLoad128(val, &input[i]);
+      streamStore128(&buffers[rank][i], val);
+    }
```
IIUC, we are loading 8 x bf16 at one time.
But here `stride` seems to be at the unit of bf16. 
Am I missing anything?

yifuwang replied:
Hi @kwen2501, the unit of `stride` is `numelPerThread = kBytesPerThread / sizeof(at::BFloat16)` which is also 8 x bf16.

yifuwang commented:
@pytorchbot merge

Commit:
Update on "Fuse the input -> p2p buffer copy into one-shot all-reduce kernel when the input is small"


This improves the gpt-fast llama2 70B 8xH100 (non-standard) TP benchmark from 86 tok/s to 88 tok/s.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "Fuse the input -> p2p buffer copy into one-shot all-reduce kernel when the input is small"


This improves the gpt-fast llama2 70B 8xH100 (non-standard) TP benchmark from 86 tok/s to 88 tok/s.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge

yifuwang commented:
@pytorchbot merge

Commit:
Update on "Fuse the input -> p2p buffer copy into one-shot all-reduce kernel when the input is small"


This improves the gpt-fast llama2 70B 8xH100 (non-standard) TP benchmark from 86 tok/s to 88 tok/s.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
